### ISSUE TITLE
Replace does not work for services

### ISSUE THREAD

**Issue by sjmiller609** [kubernetes/kubernetes#91459]
**What happened**:

[This issue on kubectl](https://github.com/kubernetes/kubectl/issues/798#issue-547658418) was closed as not a problem with kubectl, but with the server

**What you expected to happen**:

'kubectl replace' with no change on a service should work.

**How to reproduce it (as minimally and precisely as possible)**:

reproduction included in this issue https://github.com/kubernetes/kubectl/issues/798#issue-547658418

**Anything else we need to know?**:

Seems to me like this comment sums up the options https://github.com/kubernetes/kubectl/issues/798#issuecomment-577321234

I got here from this issue https://github.com/helm/helm/issues/6378#issuecomment-572716384

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by jansenmccloud** [comment]
After a long day of reading and experimenting with the Kubernetes API i found a work-around I want to share with you.

**Problem in my case:**
- I'm using the Java Client K8 API, especially `coreV1Api.replaceNamespacedService(...v1Service...)`
- I got the exact same error like written in https://github.com/kubernetes/kubectl/issues/798 and here in this issue
summary: 
I send an update of a service WITHOUT specifying the spec.clusterip within the V1Service object. The request gets then rejected with the reason spec.clusterip field is immutable and should not have the value "".

**Workaround:**
First load the existing service that contains the current clusterIP. Set the old clusterIp to the updated V1Service.

Also load the old metaData.resourceVersion value and set it to the new V1Service object, because leaving the value out using the Java API Client results in an error: `metadata.resourceVersion: Invalid value: "": must be specified for an update`

Please do not close this issue due to inactivity again, like it was done with https://github.com/kubernetes/kubectl/issues/798 .
Even if there are some workarounds, this issue will pop up again and again, because there really is a problem with the API restrictions or at least the messages you will get when interacting with the API.


**Constructive Suggestion:**
e.g. instead of throwing the error `invalid: spec.clusterIP: Invalid value: "": field is immutable` you could throw something like 

`spec.clusterIP: Invalid value: "": needs to be equal to the clusterIp of the service you want to update`

and for the resourceVersion:

`metadata.resourceVersion: Invalid value: "": must be the resourceVersion of the service you want to update`

**Comment by Nuru** [comment]
Seriously, this has been going on for years. #65272 is over 2 years old and there have been dozens of bug reports/feature requests (in Kubernetes, Kubectl, Helm, and elsewhere) showing the problems the current behavior is causing. (List available on request.)

The API should accept `clusterIP: ""` as meaning "assign a clusterIP" on creation and "keep the current clusterIP" on update. Since, when you set `clusterIP: ""` on creation, the resource does not end up with `clusterIP: ""`, this means a resource cannot validly have `clusterIP: ""`, and since the field is immutable, `clusterIP: ""` could never be a correct value for modification under the current API. 

So rejecting `clusterIP: ""` as a modification is not helping anyone. It is not rejecting an accidental change or avoiding confusion because the meaning is unclear. It is just causing problems for the sake of some kind of theoretical purity that is in the eye of the beholder. Treating `clusterIP: ""` as meaning "system should replace this with a valid setting" is equally pure.

**Comment by dims** [comment]
cc @liggitt 

long-term-issue (note to self)

**Comment by lygasch** [comment]
keepalive

**Comment by Nuru** [comment]
@dims @liggitt Is there something I can do to move this forward?

**Comment by liggitt** [comment]
/remove-sig cli
/sig network

/assign @thockin 
this is sort of the inverse of the issue fixed in https://github.com/kubernetes/kubernetes/pull/95196

@Nuru 

Using `kubectl replace` with a static manifest on an object that has spec fields populated by the API server (like clusterIP in service) or controller (like replicas in deployment) is an anti-pattern. `kubectl apply` is a better match for this usage. Alternatively, a controller that did a 'read -> modify -> replace' could be ok, since it would be starting with the current API server state.

**Comment by thockin** [comment]
This is not _impossible_ to fix in the PUT path (`kubectl replace`), but it basically turns PUT into PATCH. We have distinct PATCH operations for a reason, and that is almost certainly what you want in this case.   `kubectl apply` should "just work" as a replacement for `replace`.

Is there a reason why you can't use PATCH or apply?

@liggitt - while I continue to argue that `replace` is wrong here, maybe we should just cave and make it quietly patch the value of clusterIP in this specific case?  I really dislike it, not just for purity reasons, but because I have a hard time seeing if that will bite us later.  Thoughts?

**Comment by liggitt** [comment]
> @liggitt - while I continue to argue that replace is wrong here, maybe we should just cave and make it quietly patch the value of clusterIP in this specific case? I really dislike it, not just for purity reasons, but because I have a hard time seeing if that will bite us later. Thoughts?

I really dislike it as well. "replace all spec content" is a bad thing to teach clients to do on resources that have fields set by different actors.

**Comment by thockin** [comment]
Agree, and while we might be able to quietly patch this SPECIFIC case, we can't do that in general.

@sjmiller609 - My inclination is to close this.  I know it's not what you want, but I do think you and most users will be better served by patch or apply in almost all cases (knowing that Service has an ambiguity around changing ports that overlap each other which is not easily fixed because it is breaking).

Is there a reason why you can't use patch or apply?

**Comment by lygasch** [comment]
This really seems related to bug in Helm, as far as what I have read, and some recent (Jan2021) posts suggest that was fixed. Have not tested this out myself. Agree that adhering to expected PUT/PATCH api behavior would be best. Thanks for the followup!

**Comment by Nuru** [comment]
> Agree, and while we might be able to quietly patch this SPECIFIC case, we can't do that in general.
> 

@thockin  I am not advocating that you turn PUT into PATCH in general; I am arguing that you should honor the principle that [`PUT` is _idempotent_.](https://restfulapi.net/idempotent-rest-apis/) The definition of the `clusterIP` field in this SPECIFIC case (where the result of PUTting a value does not result in that value being set in the resource) needs be handled specially in order to fulfill the contract of `PUT`.

@lygasch I reject the idea that this is a bug in Helm, even though Helm is working hard (and failing) to handle the situation. I agree with you when you "Agree that adhering to expected PUT/PATCH api behavior would be best". My point is that the expected "PUT" behavior is that if a "PUT" request succeeds, the exact same "PUT" request given a second time should succeed as well. That is what is broken and I am asking to have fixed. 

> @sjmiller609 @Nuru @lygasch @jansenmccloud - My inclination is to close this. I know it's not what you want, but I do think you and most users will be better served by patch or apply in almost all cases (even knowing that Service has an ambiguity around changing ports that overlap each other which is not easily fixed because it is breaking).
> 
> Is there a reason why you can't use patch or apply?

Is there a better reason why I should be prohibited from using replace on a Service, when I can use it for every other resource? Why should Service require special-case handling by the user?

I ask you to think of this in the larger context of using Infrastructure as Code (IaC) to manage resources, and the automation around that. In the IaC definition of a Service, the `clusterIP` has to be set to `""` in order to get a system-assigned address. 

Semantically, `PUT` says "make the resource exactly like this". It can be, and is, used to ensure there is not "drift" in the configuration, i.e. that the configuration stored in the IaC code base is live. It is the correct thing to use when you want to say "here is the configuration that should be active", and it is what automation uses to reset the configuration after it has drifted. Users should not have to treat Services specially, they should be able to `replace` any resources with the same spec that created them.

Stating this another way, [the general contract of `PUT` is that it is _idempotent_.](https://restfulapi.net/idempotent-rest-apis/) (A user should be able to `PUT` the same thing repeatedly, successfully, and without causing changes after the first accepted `PUT`.) By not accepting `""` for `clusterIP` in `PUT`, you are breaking the contract of making `PUT` _idempotent_. 

(As a matter of practicality, the Helm maintainers are trying to make up for the fact that `PUT` is not _idempotent_, but that is not an excuse for Kubernetes to leave this API in its current state.)


As I stated above, I am not even advocating for this case to be handled as a `PATCH`. As I see it, the value of `clusterIP` is handled specially but not idempotently. The value of `""` currently means one thing if the resource does not exist and something else if it does exist. I am asking instead that it consistently, idempotently be taken as meaning "any valid address: the existing one if one is set, otherwise one randomly assigned by the controller".

**Comment by thockin** [comment]
>> Agree, and while we might be able to quietly patch this SPECIFIC case, we can't do that in general.
>
> @thockin I am not advocating that you turn PUT into PATCH in general; I am arguing that you should honor the principle that PUT is idempotent.

Unfortunately, to make PUT idempotent in this case, it would need to turn into
a variant of PATCH.  I'm not even saying I totally disagree, but that is what
it means.

As for PUT being idempotent, I don't think that's ever really been a hard
objective.  In this case (and look, I am not saying I am happy about it) that
means reading existing state and patching it into the PUT.  We *can* do that
(probably) but I can't say for sure that there are no other APIs where this
will happen (there probably are not).

Note that to do this we would need to fixup `clusterIP`, `ports[*].nodePort`,
and probably the newer dual-stack-related fields.

> @lygasch I reject the idea that this is a bug in Helm

Can someone help me understand the "bug"?  Is it about using PUT instead of
apply?

I could see two paths to fixing this, if we want to (and I am still not 100%).

1) Change clusterIP and nodePort fields to pointers.  If not-specified on PUT
AND the object exists AND the old `type` warrants a value AND the new `type`
warrants a value, copy the existing values.  This is a bit tricky for nodePort
- we'd have to at least specify what happens if the list is reordered or
  otherwise modified.  

For example, if the existing object has:

```
ports:
- name: p
  port: 80
  nodePort: 31280
  targetPort: 967
- name: q
  port 81
  nodePort: 31281
  targetPort: 5309
```

and the update specifies:

```
ports:
- name: q
  port 81
  targetPort: 5309
- name: p
  port: 80
  targetPort: 967
```

Do I track the move by name?  What if the names changed, does that lose the
tracking?  We've had innumerable issues here because the API-specified
merge-key was (is) ambiguous.

The only way to get exactly what you want in ALL cases is to RMW the resource.

Explicitly setting "" and 0 would work on initialization (compat) but would not
be auto-patched.

2) Leave the fields as non-pointers and accept "" and 0 as "DWIM" mode.  I
dislike this because it makes it less obvious when a user mistakenly tries to
clear a field - it used to fail and will now "succeed" but silently change the
value.

I am not sure I see a clever 3rd option.

The only reason I entertain this at all is because of the issues around
apply/patch and ports.

NOTE: My willingness to think it thru is not a commitment to implement it.  I
am swamped right now :)

**Comment by Nuru** [comment]
@thockin wrote
> As for PUT being idempotent, I don't think that's ever really been a hard objective.

I am shocked that you would think that, although I don't really know what you mean by "hard objective". The [Kubernetes API Concepts](https://kubernetes.io/docs/reference/using-api/api-concepts/) doc certainly implies `PUT` should be idempotent, via such statements as

> The Kubernetes API is a resource-based (RESTful) programmatic interface provided via HTTP. It supports retrieving, creating, updating, and deleting primary resources via the standard HTTP verbs (POST, PUT, PATCH, DELETE, GET)...

and 

> All objects will have a unique name to allow idempotent creation and retrieval

Moving on to how to do make it so, the idea of `PUT` is that at the end of the operation, the object is in the state described by the action. If you PUT ports P and Q and the existing object has ports R and S, you delete R and S and add P and Q. If adding or deleting a port is not allowed, then you reject the PUT. 

I think to some extent you are overthinking/overdesigning a solution here. In Kubernetes, every PUT is to some extent a patch because there are some fields that are not specifiable via PUT, like all the managed fields, or `creationTimestamp`, and fields like `selfLink` and `resourceVersion` are special in a whole different way. But that stuff should already be handled by the more generalized resource handling architecture. The general rule about PUT is that PUT on an existing resource should adjust that resource to be in the same working configuration as if you used that PUT to create a new resource. **Calling PUT a second time with the same input should be acceptable and should not change the resource**.

I believe that Kubernetes generally has a canonical ordering of all fields in a resource (and if there is no clear way to order a list, the canonical order is simply the order in which the list was originally specified on resource creation), so what you do is enumerate the fields of the existing resource in canonical order, filter out the fields like `creationTimestamp` that PUT cannot specify, and then compare it to the PUT resource, after normalizing it into canonical order, and add/delete/replace field by field.  

Clearly something like this is already happening in handling PUT and PATCH. This should not be a lot of new work.

The ongoing community complaint is that when PUT says `clusterIP: ""` and it is matched against the exiting `clusterIP: 172.20.4.5` the API rejects it, saying you can't change the `clusterIP`. All we are asking for is that this case be interpreted differently and allowed.  `clusterIP: ""` is documented as an allowable value in the input, but it is never an allowable value in the realized resource, so it cannot possibly mean to be interpreted as "force this value on the resource". Its meaning is that the system should assign a valid value and the user does not care what it is. So on PUT, when `clusterIP: ""` is matched against the exiting `clusterIP: 172.20.4.5`, he API should just confirm that `clusterIP: 172.20.4.5` is a valid value and, if so, accept the input.

**Comment by thockin** [comment]
>> As for PUT being idempotent, I don't think that's ever really been a hard objective.
>
>I am shocked that you would think that, although I don't really know what you mean by "hard objective". The [Kubernetes API Concepts](https://kubernetes.io/docs/reference/using-api/api-concepts/) doc certainly implies `PUT` should be idempotent, via such statements as
>
> `The Kubernetes API is a resource-based (RESTful) programmatic interface provided via HTTP. It supports retrieving, creating, updating, and deleting primary resources via the standard HTTP verbs (POST, PUT, PATCH, DELETE, GET)...`
>
>and 
>
>> `All objects will have a unique name to allow idempotent creation and retrieval`

What I meant is it's just never been a primary question in reviewing APIs.  Service is a weird case, far outside the norm and I'm not excusing it - just explaining what I meant. 

>Moving on to how to do make it so, the idea of `PUT` is that at the end of the operation, the object is in the state described by the action. If you PUT ports P and Q and the existing object has ports R and S, you delete R and S and add P and Q. If adding or deleting a port is not allowed, then you reject the PUT. 

For historical reasons, the "key" for ports was not the name but the port value, then the value + protocol.  So we can key off that, I think.

>I think to some extent you are overthinking/overdesigning a solution here. In Kubernetes, every PUT is to some extent a patch because there are some fields that are not specifiable via PUT, like all the managed fields, or `creationTimestamp`, and fields like `selfLink` and `resourceVersion` are special in a whole different way. 

Those are all outside of `spec`.  I don't know of any resource that PUT-PATCH inside the spec.

> **Calling PUT a second time with the same input should be acceptable and should not change the resource**.

Hilariously, I just noticed that nodePort is even worse - it accepts the non-value and then re-allocates the nodePort.  I am going to call this a bug, but now we have a compatibility discussion to have.  If a user does not specify a value, we can reasoinably argue that we should retain existing values.  But if they specify `0`, do we stay compatible and re-allocate the node port, or do we say that was always a bug?

Is it a useful "feature" to be able to re-allocate?  We disallow it explicitly gfor ClusterIPs.

>Clearly something like this is already happening in handling PUT and PATCH. This should not be a lot of new work.

You haven't spent much time in the Service REST code, I guess.  I have a large rework of that on-going, but this was not part of it.  :)

>The ongoing community complaint is that when PUT says `clusterIP: ""` and it is matched against the exiting `clusterIP: 172.20.4.5` the API rejects it, saying you can't change the `clusterIP`.

Given the previous point about nodePort, would it be fair to modify your statement as:

"when PUT says `clusterIP: null` (or does not specify at all) ..."

With the corollart that if you actually specify `clusterIP: ""` you will get the current (immutable error) behavior.

This seems like it would satisfy helm and other users who are not specifying literal "" - they are not specifying any value at all.

Thoughts?

**Comment by thockin** [comment]
At this point I am willing to call it a bug, albeit an unpleasant one.

**Comment by Nuru** [comment]
@thockin wrote

> Given the previous point about nodePort, would it be fair to modify your statement as:
> 
> "when PUT says `clusterIP: null` (or does not specify at all) ..."
> 
> With the corollart that if you actually specify `clusterIP: ""` you will get the current (immutable error) behavior.
> 
> This seems like it would satisfy helm and other users who are not specifying literal "" - they are not specifying any value at all.
> 

No, not at all. The user community, including Helm maintainers and especially ingress controller maintainers, went to a lot of effort to work around what I am calling a bug (where `clusterIP: ""` was rejected) by going out of their way to not specify `clusterIP` at all so that at least `PATCH` would work. I view that as a hack to work around a bug, not a proper state of affairs. (They took advantage of the fact that `clusterIP: ""` was the default, but if it were not the default, then we would be left with no way to even PATCH a service using its initial spec.) I maintain my position that specifying `clusterIP: ""`, which is explicitly allowed in the API for creating the service, should continue to be allowed in any `PUT` (or `PATCH` for that matter). 

I have had regular run-ins with this `clusterIP` issue for years, which is why I'm being insistent on getting it fixed properly. I do not recall having similar problems with `nodePort` and so I am less informed on those issues. I don't know what `nodePort: 0` means in the spec or in practice, so I don't have a specific opinion as to what it should do to an existing resource. I will simply restate the general principle which should apply to `nodePort` just like everything else: what ever you allow as a setting on creation should be allowed in any `PUT` and should not cause a change if repeatedly applied. 

The very practical reason for this is so that I can record some place (like a `service.yaml` checked into a VCS) that "this is how the resource should be configured" and use that (1) to create the resource and (2) to return the resource to expected configuration after some external operation changed it, using a standard API while otherwise treating the resource as opaque.

**Comment by khenidak** [comment]
It is a bug. Not because it does not match rest api semantics. But because it is not consistent with the rest of the apis. In other words, If user can do `replace (aka idempotent PUT)` on other resources they should be able to do it on `Services` as well. It is hard (and weird) to ask users to do one thing for all resources, but then do a totally different thing for Services.

As discussed over sig-network the problem will be user intent. Some ideas for a solution:

- If `old.Spec.ClusterIP=<IP> && new.Spec.ClusterIP=<empty-string>` irrespective of changes in other `new.Spec.OtherFields` we can assume that the user is replacing and does not care about ClusterIP value, hence copying `old.Spec.ClusterIP` to new `new.Spec.ClusterIP` should be fine. It is basically like the user saying `I am ok with whatever clusterIP you give me, or change it` so copying old=>new should be Ok.
- If `old.Spec.ClusterIP=<IP> && new.Spec.ClusterIP=<other-IP>` that is a failure case, because it breaks ClusterIP immutability rules (@thockin we should really, really remove ClusterIP immutability but i digress).
- We will still require the validation to work as is. e.g. replacing with `service.Spec.Type` with a type that does not require ClusterIP and not setting it to `<empty-string>` should fail.

I can look into building a POC for this.

**Comment by jglick** [comment]
I am probably missing some subtlety here, but as in https://github.com/helm/helm/issues/7956#issuecomment-640912141 I suspect that Helm is behaving properly and the problem is in way Kubernetes handles `Service`s: the `spec` of a resource should never be rewritten with variable environmental information; that is what `status` is for. In other words `.spec.clusterIP` could be set in the unusual case that you wished to pick an IP yourself but would normally be null; `.status.clusterIP` would reflect the service’s current value, expected to be retained across unrelated `.spec` patches.

**Comment by bacongobbler** [comment]
@khenidak nailed it on the head with his last comment. I do agree that we should not be re-defining the behaviour of a PUT request. But the inconsistency of how a Service updates the `spec.clusterIP` field compared to other resources results in inconsistencies in the PUT behaviour.

Once a resource defines fields in the `spec` object, normally Kubernetes leaves it unchanged. My understanding is that Kubernetes uses this `spec` field to use as a reference point to be the "ideal state of the world", using the `status` field to record the "current state of the resource".

The only offending case I've seen in the wild is a `Service`, as it updates both the `status` field AND the `spec.ClusterIP` field. The end result is that a `kubectl replace` with a `Service` defining `spec.ClusterIP = ""` results in differing behaviour than the rest of the resource APIs.

From Helm's point of view, a chart author's only course of action is to

a) omit `spec.clusterIP` from the manifest and all subsequent updates, or
b) if you happened to define `spec.clusterIP = ""` in your first install, ensure it is NOT present in subsequent updates so that the value Kubernetes provides is preseved.

Failing to do so results in a failed PUT (`helm upgrade --force`) request. I wrote this behaviour down in-depth in this Helm comment: https://github.com/helm/helm/issues/6378#issuecomment-557746499

I'd propose an alternative solution than @khenidak's solutions: instead of updating `spec.clusterIP` after an IP has been assigned to the Service, use the `status` API, i.e. `status.clusterIP`. It would be a breaking change to the API, but it should bring Services back in line with the rest of the resource APIs, leaving `spec.clusterIP` as a user-defined field.

**Comment by cndoit18** [comment]
> @khenidak nailed it on the head with his last comment. I do agree that we should not be re-defining the behaviour of a PUT request. But the inconsistency of how a Service updates the `spec.clusterIP` field compared to other resources results in inconsistencies in the PUT behaviour.
> 
> Once a resource defines fields in the `spec` object, normally Kubernetes leaves it unchanged. My understanding is that Kubernetes uses this `spec` field to use as a reference point to be the "ideal state of the world", using the `status` field to record the "current state of the resource".
> 
> The only offending case I've seen in the wild is a `Service`, as it updates both the `status` field AND the `spec.ClusterIP` field. The end result is that a `kubectl replace` with a `Service` defining `spec.ClusterIP = ""` results in differing behaviour than the rest of the resource APIs.
> 
> From Helm's point of view, a chart author's only course of action is to
> 
> a) omit `spec.clusterIP` from the manifest and all subsequent updates, or
> b) if you happened to define `spec.clusterIP = ""` in your first install, find the value Kubernetes provided and inject it into your template so that the value Kubernetes provided is preseved.
> 
> Failing to do so results in a failed PUT (`helm upgrade --force`) request. I wrote this behaviour down in-depth in this Helm comment: [helm/helm#6378 (comment)](https://github.com/helm/helm/issues/6378#issuecomment-557746499)
> 
> I'd propose an alternative solution than @khenidak's solutions: instead of updating `spec.clusterIP` after an IP has been assigned to the Service, use the `status` API, i.e. `status.clusterIP`. It would be a breaking change to the API, but it should bring Services back in line with the rest of the resource APIs, leaving `spec.clusterIP` as a user-defined field.

I agree 👍

**Comment by aojea** [comment]
> I'd propose an alternative solution than @khenidak's solutions: instead of updating `spec.clusterIP` after an IP has been assigned to the Service, use the `status` API, i.e. `status.clusterIP`. It would be a breaking change to the API, but it should bring Services back in line with the rest of the resource APIs, leaving `spec.clusterIP` as a user-defined field.

I'm afraid that breaking an API is not a feasible option https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md

**Comment by bacongobbler** [comment]
That document does allow for breaking changes under the apiVersion. Instead of making a breaking change to v1 services, you can introduce Service v2. Existing v1 services migrate their spec.clusterIP’s existing value (auto-generated or not) to the new version. spec.clusterIP and status.clusterIP would now match their corresponding relationships.

**Comment by sftim** [comment]
(If it's not clear whether the API changes document does or doesn't allow the change from @bacongobbler's last comment, let's raise a separate improvement request against that document).

**Comment by cabrinoob** [comment]
Does the case is important ? Because here I can read `clusterIP` or `ClusterIP` ...

**Comment by thockin** [comment]
I have a patch.  I'll try to get it out for review tonight.

**Comment by jglick** [comment]
Has anyone yet checked that the purported fix actually addresses the issue in Helm? https://github.com/helm/helm/issues/6378#issuecomment-850530229

**Comment by thockin** [comment]
Please do!

**Comment by jwh-hutchison** [comment]
> Has anyone yet checked that the purported fix actually addresses the issue in Helm? [helm/helm#6378 (comment)](https://github.com/helm/helm/issues/6378#issuecomment-850530229)

I can still see the error but I'm on GKE 1.18, will this only apply to Kubernetes 1.21/22?

**Comment by aojea** [comment]
> I can still see the error but I'm on GKE 1.18, will this only apply to Kubernetes 1.21/22?

It will be part of 1.22 or the next 1.21.X (I don't remember now what is the latest stable)

**Comment by denispilon** [comment]
Can anyone confirm that this issue was actually fixed in k8s version 1.21.4 or 1.22.1 ?
I do not see it listed in the release notes/change log

**Comment by bacongobbler** [comment]
The change was released in the 1.22.0-beta.2, so yes it should be available in 1.22. See https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.22.md#api-change-2

I do not believe it was backported to 1.21.

**Comment by denispilon** [comment]
> The change was released in the 1.22.0-beta.2, so yes it should be available in 1.22. See https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.22.md#api-change-2
> 
> I do not believe it was backported to 1.21.

Thanks for confirming and providing the details @bacongobbler
